File: python/pid.py
class PID:
    def __init__ (self,  kp, ki, kd, setpoint = 0.0, integral_limit = None):
    
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.setpoint = setpoint
        self.integral = 0.0
        self.last_measurement = 0.0
        self.integral_limit = integral_limit


    def compute(self, mpu_reading, dt):

        error = self.setpoint - mpu_reading
        self.integral += error *dt
        if self.integral_limit is not None:
            if self.integral > self.integral_limit:
                self.integral = self.integral_limit
            elif self.integral < -self.integral_limit:
                self.integral = -self.integral_limit
        derivative = -(mpu_reading - self.last_measurement) / dt
        self.last_measurement = mpu_reading
        return self.kp * error + self.ki * self.integral + self.kd * derivative
    
    def reset(self):
        self.integral = 0.0
        self.last_measurement = 0.0

File: firmware/balance/PID.cpp

#include "PID.h"


PID::PID(float p, float i, float d)
{
    kp = p;
    ki = i;
    kd = d;
    integral = 0;
    prevMeasurement = 0;
    firstCall = true;

    if (ki == 0)
    {
        integralMax = 0;
    }
    
    else
    {
        integralMax = 255 /ki;
    }
    
}

float PID::update(float measurement, float setpoint, float dt)
{
    float error = setpoint - measurement;

    integral += error * dt;

    if (integral > integralMax)
    {
        integral = integralMax;
    
    }
    else if (integral < -integralMax)
    {
        integral = -integralMax;
    }
    float derivative;

    if (firstCall == true)
    {
        derivative = 0;
        firstCall = false;
    }
    else 
    {
        if (dt == 0)
        {
            derivative = 0;
        }
        else
        {
            derivative = -kd * (measurement - prevMeasurement) / dt;
        }

    }

    prevMeasurement = measurement;

    return kp*error + ki*integral + derivative;
}

void PID::reset()
{
    integral = 0;
    prevMeasurement = 0;
    firstCall = true;

}

void PID :: setGains( float p, float i, float d)
{
    kp = p;
    ki = i;
    kd = d;
    if (ki == 0)
    {
        integralMax = 0;
    }
    
    else
    {
        integralMax = 255 /ki;
    }
    
}